File: ifaddrs4j/source/production/java/io/oddsource/java/net/ifaddrs4j/InterfaceBrowser.java
/*
 * Copyright © 2010-2023 OddSource Code (license@oddsource.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.oddsource.java.net.ifaddrs4j;

import java.lang.ref.Cleaner;
import java.util.Enumeration;
import java.util.Iterator;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Stream;

/**
 * A tool for browsing network interfaces/adapters configured on this system. To ensure memory resources are cleaned
 * up properly, be sure to {@link #close()} the browser when you are done with it, such as with a try-with-resources
 * block. Example uses are detailed below.<br>
 * <br>
 * Iterate over the browser to inspect all interfaces:
 * <pre>{@code
 * try(var browser = new InterfaceBrowser())
 * {
 *     ...
 *     for(var anInterface : browser)
 *     {
 *         System.out.println(anInterface.toString());
 *     }
 *     ...
 * }
 * }</pre>
 * For-each all interfaces with a lambda, stopping iteration when the lambda returns {@code false}:
 * <pre>{@code
 * try(var browser = new InterfaceBrowser())
 * {
 *     ...
 *     var notFound = browser.forEachInterface(iface -> {
 *         if(... some test against iface ...)
 *         {
 *             ... // use the iface
 *             return false; // to stop iteration
 *         }
 *         return true; // to continue iteration
 *     });
 *     if(notFound)
 *     {
 *         ... // react to this situation
 *     }
 * }
 * }</pre>
 * Other actions you can take, all of which should be surrounded with try-with-resources:
 * <pre>{@code
 * // get an interface if you know its system name (on Windows, you can use its
 * // UUID string or friendly name; either will match)
 * var iface = browser.getInterface("eth0");
 *
 * // get an interface if you know its system index
 * var iface = browser.getInterface(1);
 *
 * // get an enumeration of all interfaces
 * var interfaces = browser.getInterfaces();
 *
 * // stream over the interfaces, finding the first non-loopback, status-up interface
 * // with at least one non-link-local IPv6 address
 * var iface = browser.interfaces().
 *                     filter(i -> i.isFlagEnabled(InterfaceFlag.IsUp) &&
 *                                 !i.isFlagEnabled(InterfaceFlag.IsLoopback)).
 *                     findFirst(i -> i.ipv6Addresses().findFirst(
 *                         a -> !a.getAddress().isLinkLocalAddress()
 *                     ) != null);
 * }</pre>
 * @implNote
 *     This browser's data does not become "exhausted," though it could become stale. Upon first post-construction
 *     use of {@link #interfaces()}, {@link #iterator()}, {@link #forEachInterface(Function)},
 *     {@link #getInterfaces()}, {@link #getInterface(String)}, or {@link #getInterface(int)}, the browser holds
 *     a snapshot-in-time of the system's network interface configuration, which may immediately become stale. These
 *     six accessor methods can be used any number of times, in any order, in any combination, and by multiple
 *     threads if necessary.
 *
 * @since 1.0.0
 */
public final class InterfaceBrowser implements AutoCloseable, Iterable<Interface>
{
    static
    {
        if(!Loader.LOADED)
        {
            // this will never actually happen, but we need to ensure that Loader initializes
            throw new RuntimeException("Not possible!");
        }
    }

    private static final Cleaner cleaner = Cleaner.create();

    private final NativeBrowser internal;

    private final Cleaner.Cleanable cleanable;

    /**
     * Construct an interface browser.
     */
    public InterfaceBrowser()
    {
        this.internal = new NativeBrowser();
        this.cleanable = InterfaceBrowser.cleaner.register(this, new BrowserCleaner(this.internal));
    }

    @Override
    public void close()
    {
        this.cleanable.clean();
    }

    private synchronized void ready()
    {
        if(this.internal == null || this.internal.isClosed())
        {
            throw new IllegalStateException("This interface browser has been closed or is in an unusable state.");
        }
    }

    /**
     * Get the interface with the given name, or {@code null} if no such interface exists. On Windows,
     * an attempt will be made to match the given name against the friendly name, the UUID, and the
     * UUID with curly braces around it.<br>
     *
     * @implNote
     *     The complexity of this method is {@code O(n)} on the first call to any of the six accessor
     *     methods on this instance and then {@code O(1)} for all subsequent calls.
     *
     * @param name The interface name or, optionally on Windows (only), the UUID or <code>{UUID}</code>
     * @return the matched interface or {@code null}.
     */
    public Interface getInterface(final String name)
    {
        this.ready();
        return this.internal.getInterface(name);
    }

    /**
     * Get the interface with the given index, or {@code null} if no such interface exists.<br>
     *
     * @implNote
     *     The complexity of this method is {@code O(n)} on the first call to any of the six accessor
     *     methods on this instance and then {@code O(1)} for all subsequent calls.
     *
     * @param index The interface index
     * @return the matched interface or {@code null}.
     */
    public Interface getInterface(final int index)
    {
        this.ready();
        return this.internal.getInterface(index);
    }

    /**
     * Get an iterable enumeration of all interfaces configured on the system.<br>
     * <br>
     * Note that there is no guaranteed order in which interfaces will be enumerated, and that the
     * order can change from one call to the next (though that would not be typical). If you need
     * particular order, consider Streaming {@link #interfaces()} and sorting the stream.
     *
     * @return an enumeration of interfaces.
     */
    public Enumeration<Interface> getInterfaces()
    {
        this.ready();
        return new InterfaceEnumeration(this.internal.getInterfaces().listIterator());
    }

    /**
     * Get an unmodifiable iterator of all interfaces configured on the system, which will throw
     * {@link UnsupportedOperationException} upon an attempt to call {@link Iterator#remove()}.<br>
     * <br>
     * Note that there is no guaranteed order in which interfaces will be iterated, and that the
     * order can change from one call to the next (though that would not be typical). If you need
     * a particular order, consider Streaming {@link #interfaces()} and sorting the stream.
     *
     * @return an iterator of interfaces.
     */
    @Override
    public Iterator<Interface> iterator()
    {
        this.ready();
        return new InterfaceIterator(this.internal.getInterfaces().listIterator());
    }

    /**
     * Get a filterable stream of all interfaces configured on the system.<br>
     *
     * @implNote
     *     That there is no guaranteed order in which interfaces will be streamed unless you
     *     {@link Stream#sorted(java.util.Comparator) sort} the stream. The only guarantee is that subsequent streams
     *     will always be in the same order, and that streaming, {@link #iterator() iterating},
     *     {@link #getInterfaces() enumerating}, and {@link #forEachInterface(Function) for-each-ing} will occur in
     *     the same order.
     *
     * @return a stream of interfaces.
     */
    public Stream<Interface> interfaces()
    {
        this.ready();
        return this.internal.getInterfaces().stream();
    }

    /**
     * For each interface configured on the system, apply the supplied function, stopping when
     * the function returns {@code false}.<br>
     * <br>
     * Note that there is no guaranteed order in which interfaces will be iterated, and that the
     * order can change from one call to the next (though that would not be typical). If you need
     * a particular order, consider Streaming {@link #interfaces()} and sorting the stream.
     *
     * @param doThis The function to apply to each interface
     * @return {@code true} if the function was applied to all interfaces, {@code false} if the function
     *         ever returned false to stop application. As an edge case, this may return {@code false}
     *         if the function was applied to all interfaces but the function returned {@code false} on
     *         the very last interface.
     */
    public boolean forEachInterface(final Function<Interface, Boolean> doThis)
    {
        this.ready();
        return this.internal.forEachInterface(doThis);
    }

    @Override
    public String toString()
    {
        final var builder = new StringBuilder("<io.oddsource.java.net.ifaddrs4j.InterfaceBrowser: ");
        if (this.internal == null)
        {
            builder.append("uninitialized/unusable");
        }
        else if(this.internal.isClosed())
        {
            builder.append("closed/unusable");
        }
        else if(this.internal.pBrowser == 0)
        {
            builder.append("internally uninitialized/unusable");
        }
        else if(this.internal.interfaces == null)
        {
            builder.append("pending population, internal pointer = 0x").
                    append(Long.toHexString(this.internal.pBrowser));
        }
        else
        {
            builder.append("contains info on ").append(this.internal.interfaces.size()).append(" interfaces").
                    append(", internal pointer = 0x").append(Long.toHexString(this.internal.pBrowser));
        }
        builder.append(">");
        return builder.toString();
    }

    private static final class InterfaceIterator implements Iterator<Interface>
    {
        private final Iterator<Interface> iterator;

        InterfaceIterator(final Iterator<Interface> iterator)
        {
            this.iterator = iterator;
        }

        @Override
        public boolean hasNext()
        {
            return this.iterator.hasNext();
        }

        @Override
        public Interface next()
        {
            return this.iterator.next();
        }
    }

    private static final class InterfaceEnumeration implements Enumeration<Interface>
    {
        private final Iterator<Interface> iterator;

        InterfaceEnumeration(final Iterator<Interface> iterator)
        {
            this.iterator = iterator;
        }

        @Override
        public boolean hasMoreElements()
        {
            return this.iterator.hasNext();
        }

        @Override
        public Interface nextElement()
        {
            return this.iterator.next();
        }
    }

    private static final class NativeBrowser implements AutoCloseable
    {
        /**
         * Stores the C++ native pointer to the OddSource::Interfaces::InterfaceBrowser object, used
         * internally by native code.
         */
        @SuppressWarnings("unused") // it is used by native code
        private long pBrowser;

        /**
         * Stores the cached Java conversions of the native interface info, set internally by native code.
         */
        @SuppressWarnings("unused") // it is used by native code
        private List<Interface> interfaces;

        private boolean closed;

        NativeBrowser()
        {
            this.init();
        }

        private synchronized native void init();

        public synchronized boolean isClosed()
        {
            return this.closed;
        }

        @Override
        public synchronized void close()
        {
            if (!this.closed)
            {
                this.closed = true;
                this.closeInternal();
            }
        }

        private native void closeInternal();

        synchronized native Interface getInterface(String name);

        synchronized native Interface getInterface(int index);

        synchronized List<Interface> getInterfaces()
        {
            this.ensureInterfacesPopulated();
            return this.interfaces;
        }

        private synchronized native void ensureInterfacesPopulated();

        synchronized boolean forEachInterface(final Function<Interface, Boolean> doThis)
        {
            if (this.interfaces != null)
            {
                for (final var iface : this.interfaces)
                {
                    if (!doThis.apply(iface))
                    {
                        return false;
                    }
                }
                return true;
            }

            return this.forEachInterfaceAndPopulate(doThis);
        }

        synchronized native boolean forEachInterfaceAndPopulate(Function<Interface, Boolean> doThis);
    }

    private static final class BrowserCleaner implements Runnable
    {
        private final NativeBrowser subject;

        BrowserCleaner(final NativeBrowser subject)
        {
            this.subject = subject;
        }

        @Override
        public void run()
        {
            this.subject.close();
        }
    }
}

File: ifaddrs4j/source/production/cpp/InterfaceBrowser.cpp
/*
 * Copyright © 2010-2023 OddSource Code (license@oddsource.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "generated/io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser.h"

#include <ifaddrs4cpp/Interfaces.h>

#include "cache.h"
#include "Interface.h"
#include "IpAddress.h"
#include "MacAddress.h"
#include "macros.h"
#include "wrappers.h"

namespace
{
    using namespace OddSource::Interfaces;
    using namespace OddSource::ifaddrs4j;

    jfieldID get_field(JNIEnv * env, jobject self, char const * name, char const * signature)
    {
        auto cls(env->GetObjectClass(self));
        IF_NULL_RETURN_NULL(cls)

        return env->GetFieldID(cls, name, signature);
    }

    jfieldID get_pBrowser_field(JNIEnv * env, jobject self)
    {
        return get_field(env, self, "pBrowser", "J");
    }

    jfieldID get_interfaces_field(JNIEnv * env, jobject self)
    {
        return get_field(env, self, "interfaces", "Ljava/util/List;");
    }

    InterfaceBrowser * get_pBrowser(JNIEnv * env, jobject self)
    {
        auto field(get_pBrowser_field(env, self));
        IF_NULL_RETURN_NULL(field)

        jlong pointer(env->GetLongField(self, field));
        if (pointer > -2 && pointer < 2)
        {
            if (env->ExceptionOccurred() == NULL)
            {
                auto IllegalStateException(JCache::c(env, "IllegalStateException"s));
                IF_NULL_RETURN_NULL(IllegalStateException)
                env->ThrowNew(
                    IllegalStateException,
                    "Internal browser pointer is not set, browser has not been initialized.");
            }
            return NULL;
        }

        // This is *waves hands* magic *waves hands* to retrieve the pointer from the Java object
        auto u_pointer(static_cast<uint64_t>(pointer));
        auto browser(reinterpret_cast<InterfaceBrowser *>(u_pointer));
        return browser;
    }

    jobject get_interfaces(JNIEnv * env, jobject self)
    {
        auto field(get_interfaces_field(env, self));
        IF_NULL_RETURN_NULL(field)

        return env->GetObjectField(self, field);
    }

    jobject get_interfaces_list(JNIEnv * env, InterfaceBrowser * browser)
    {
        try
        {
            auto interfaces(browser->get_interfaces());

            ArrayList interfaces_list(env, (jint) interfaces.size());

            for (auto const & iface : interfaces)
            {
                auto jInterface(convert_to_java(env, iface));
                IF_NULL_RETURN_NULL(jInterface)
                if (!interfaces_list.add(jInterface))
                {
                    return NULL;
                }
            }

            return interfaces_list.unwrap();
        }
        CATCH_STD_EXCEPTION_THROW_EXCEPTION_IF_NOT_THROWN("RuntimeException"s, return NULL)
    }
}

/*
 * Class:     io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser
 * Method:    init
 * Signature: ()V
 */
void JNICALL
Java_io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_00024NativeBrowser_init(
    JNIEnv * env,
    jobject self)
{
    using namespace OddSource::ifaddrs4j;

    auto field(get_pBrowser_field(env, self));
    IF_NULL_RETURN_VOID(field)

    jlong pointer(env->GetLongField(self, field));
    if (pointer != 0)
    {
        auto IllegalStateException(JCache::c(env, "IllegalStateException"s));
        IF_NULL_STMT(IllegalStateException, return)
        env->ThrowNew(
            IllegalStateException,
            "Duplicate call to init(): Internal browser pointer is already set, browser was already initialized.");
        return;
    }

    try
    {
        // This is *waves hands* magic *waves hands* to store the pointer in the Java object
        auto browser(new OddSource::Interfaces::InterfaceBrowser());
        auto u_pointer(reinterpret_cast<uint64_t>(browser));
        pointer = static_cast<int64_t>(u_pointer);
    }
    CATCH_STD_EXCEPTION_THROW_EXCEPTION_IF_NOT_THROWN("RuntimeException"s, return)

    env->SetLongField(self, field, pointer);
}

/*
 * Class:     io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser
 * Method:    closeInternal
 * Signature: ()V
 */
void JNICALL
Java_io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_00024NativeBrowser_closeInternal(
    JNIEnv * env,
    jobject self)
{
    auto browser(get_pBrowser(env, self));
    IF_NULL_RETURN_VOID(browser)

    delete browser;
}

#define ENSURE_POPULATED(stmt) if (get_interfaces(env, self) == NULL) \
    { \
        if (env->ExceptionOccurred() != NULL) \
        { \
            stmt; \
        } \
        auto interfaces(get_interfaces_list(env, browser)); \
        IF_NULL_STMT(interfaces, stmt) \
        auto field(get_interfaces_field(env, self)); \
        IF_NULL_STMT(field, stmt) \
        env->SetObjectField(self, field, interfaces); \
        if (env->ExceptionOccurred() != NULL) \
        { \
            stmt; \
        } \
    }

/*
 * Class:     io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser
 * Method:    getInterface
 * Signature: (Ljava/lang/String;)Lio/oddsource/java/net/ifaddrs4j/Interface;
 */
jobject JNICALL
Java_io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_00024NativeBrowser_getInterface__Ljava_lang_String_2(
    JNIEnv * env,
    jobject self,
    jstring name)
{
    using namespace OddSource::ifaddrs4j;

    auto browser(get_pBrowser(env, self));
    IF_NULL_RETURN_NULL(browser)
    ENSURE_POPULATED(return NULL)

    char const * name_c_str(env->GetStringUTFChars(name, 0));
    IF_NULL_RETURN_NULL(name_c_str)

    ::std::string name_str(name_c_str);
    env->ReleaseStringUTFChars(name, name_c_str);

    try
    {
        return convert_to_java(env, browser->get_interface(name_str));
    }
    catch (::std::invalid_argument const &)
    {
        return NULL;
    }
    CATCH_STD_EXCEPTION_THROW_EXCEPTION_IF_NOT_THROWN("RuntimeException"s, return NULL)
}

/*
 * Class:     io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser
 * Method:    getInterface
 * Signature: (I)Lio/oddsource/java/net/ifaddrs4j/Interface;
 */
jobject JNICALL
Java_io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_00024NativeBrowser_getInterface__I(
    JNIEnv * env,
    jobject self,
    jint index)
{
    using namespace OddSource::ifaddrs4j;

    auto browser(get_pBrowser(env, self));
    IF_NULL_RETURN_NULL(browser)
    ENSURE_POPULATED(return NULL)

    try
    {
        return convert_to_java(env, browser->get_interface(index));
    }
    catch (::std::invalid_argument const &)
    {
        return NULL;
    }
    CATCH_STD_EXCEPTION_THROW_EXCEPTION_IF_NOT_THROWN("RuntimeException"s, return NULL)
}

/*
 * Class:     io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser
 * Method:    ensureInterfacesPopulated
 * Signature: ()V
 */
void JNICALL
Java_io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_00024NativeBrowser_ensureInterfacesPopulated(
    JNIEnv * env,
    jobject self)
{
    auto browser(get_pBrowser(env, self));
    IF_NULL_RETURN_VOID(browser)
    ENSURE_POPULATED(return)
}

/*
 * Class:     io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_NativeBrowser
 * Method:    forEachInterfaceAndPopulate
 * Signature: (Ljava/util/function/Function;)Z
 */
jboolean JNICALL
Java_io_oddsource_java_net_ifaddrs4j_InterfaceBrowser_00024NativeBrowser_forEachInterfaceAndPopulate(
    JNIEnv * env,
    jobject self,
    jobject function)
{
    using namespace OddSource::Interfaces;
    using namespace OddSource::ifaddrs4j;

    auto browser(get_pBrowser(env, self));
    IF_NULL_RETURN_FALSE(browser)

    Function func(env, function);
    IF_INVALID_STMT(func, return false)

    BooleanUnboxer unboxer(env);
    IF_INVALID_STMT(unboxer, return false)

    ArrayList interfaces_list(env);
    IF_INVALID_STMT(interfaces_list, return false)

    bool keep_calling_callable(true);
    bool return_error(false);

    ::std::function<bool(Interface const &)> do_this = [
        env,
        &unboxer,
        &func,
        &interfaces_list,
        &keep_calling_callable,
        &return_error](auto iface)
    {
        auto jInterface(convert_to_java(env, iface));
        if (jInterface == NULL)
        {
            return_error = true;
            return false;
        }
        if (!interfaces_list.add(jInterface))
        {
            return_error = true;
            return false;
        }

        if (keep_calling_callable)
        {
            auto result(func(jInterface));
            if (result == NULL)
            {
                if (env->ExceptionOccurred() == NULL)
                {
                    auto IllegalArgumentException(JCache::c(env, "IllegalArgumentException"));
                    if (IllegalArgumentException != NULL)
                    {
                        env->ThrowNew(
                            IllegalArgumentException,
                            "Function<Interface, Boolean> must return a boolean, not null.");
                    }
                }
                return_error = true;
                return false;
            }
            keep_calling_callable = unboxer(result);
            if (!keep_calling_callable && env->ExceptionOccurred() != NULL)
            {
                return_error = true;
                return false;
            }
        }

        return true;
    };

    try
    {
        browser->for_each_interface(do_this);
    }
    CATCH_STD_EXCEPTION_THROW_EXCEPTION_IF_NOT_THROWN("RuntimeException"s, return false)

    if (!return_error && get_interfaces(env, self) == NULL && env->ExceptionOccurred() == NULL)
    {
        auto field(get_interfaces_field(env, self));
        IF_NULL_RETURN_FALSE(field)
        env->SetObjectField(self, field, interfaces_list.unwrap());
        if (env->ExceptionOccurred() != NULL)
        {
            return false;
        }
    }

    return return_error ? false : keep_calling_callable;
}
